### ISSUE TITLE
Generalized Least Square fitting updates

### ISSUE THREAD

**Issue by karllark** [spacetelescope/jwst#2821]
Generalized Least Square (GLS) fitting is an optimal pipeline update that would be useful for both better fitting of ramps using a more complete noise model (beyond  just photon and read noise) and for better cosmic ray jump detection (CRFIT) in the read noise dominated/influenced regime.  Seems like a good idea to setup tests as a start, especially as there already is GLS code in the current pipeline.

This issue is just to say that I've started doing this on my fork of the pipeline.  Comments?  Suggestions?

**Comment by karllark** [comment]
Definitely for the future.  I see this as a long lead item for the optimal pipeline and have been thinking about it.  This is the start of working to define what is needed via CI tests.  And maybe some work to speed up the algorithm (TBD).

**Comment by stscicrawford** [comment]
@karllark I think this is already a duplicate of this ticket: https://jira.stsci.edu/browse/JP-267

But I'll leave connect these two tickets so that work is tracked, but discussion should probably be on the Jira ticket.

**Comment by karllark** [comment]
That JIRA ticket is for the whole thing, both GLS for fitting to enable better uncertainty use *and* using the CRFIT method for sensitive cosmic ray detection.

Is there a github issue for that JIRA ticket?  Or only the JIRA ticket?

**Comment by philhodge** [comment]
Are there cosmic-ray jumps in your test image?

**Comment by dmggh** [comment]
Karl -

It would be interesting to see how it performs on a large dataset having cosmic rays. If you need one I can supply it.

-Dave

________________________________

**Comment by karllark** [comment]
I agree.  I am slowly working my way to doing such a test.  I can easily generate such myself, so no worries.  I am working on this slowly, so don't expect progress quickly.  :-)

**Comment by karllark** [comment]
Results from my weekly 2 hour session on this topic (plus a bit more time this morning).

Worked through the gls_fit function.  Simplified the code a little bit, removed the use of residuals from creation of the ramp covariance matrix (not needed actually), and made a few variable changes to make it easier to understand. 

Simple timing shows that GLS takes longer for longer ramps.  This is expected given that the covariance matrix scales as (# groups)^2.   Such simple testing shows that GLS scales worse with # groups than OLS.

GLS does scales better with adding a ramp jump given that this only adds 1 parameter to the solution (i.e., slightly larger matrices) resulting is not many more calculations. 

Next steps are to implement more testing of the GLS ramp fitting method to make sure it is producing the correct results!  And investigate if speeding up the GLS calculation is possible.  Profiling shows that the majority of the GLS time is spent inverting matrices (no surprise).

**Comment by karllark** [comment]
Basic timing info.  Clear over estimate compared pytest profiling.  Need to get a better way of automated timing.  

Info is 
timing: method, # groups, CR group, time (in sec).
 
timing:  OLS 10 None 5.593559982
timing:  OLS 10 4 11.729031102999999
timing:  OLS 20 None 8.058189907000003
timing:  OLS 20 4 15.269179853
timing:  OLS 40 None 14.027408311999999
timing:  OLS 40 4 24.403618749000003
timing:  GLS 10 None 6.996902560000009
timing:  GLS 10 4 7.713892958000002
timing:  GLS 20 None 30.440593038999992
timing:  GLS 20 4 33.92807330800001
timing:  GLS 40 None 226.46407881399998
timing:  GLS 40 4 264.696551341

**Comment by melanieclarke** [comment]
Superseded by [JP-3576](https://jira.stsci.edu/browse/JP-3576) and related tickets, closing.
